src/osd/OSDMap.cc

```diff
@@ -5987,7 +5987,26 @@ void OSDMap::check_health(CephContext *cct,
   }
 
   // OSD_UPGRADE_FINISHED
-  // none of these (yet) since we don't run until luminous upgrade is done.
```

Comment:
check_health() is extremely long and ungainly. It is just a matter of taste, but
you might want to consider (as you are already changing these lines) breaking it
up with something like this:

(also - this way there is no creation of an empty require_release just before assigning to it)

```
// name the (highest) version supported by all up OSDs, if
// 'require_osd_release' was set lower than that
std::optional<std::string> OSDMap::unrequested_release(ceph_release_t requested)
{
  if (HAVE_FEATURE(get_up_osd_features(), SERVER_QUINCY) &&
      requested < ceph_release_t::quincy) {
    return "quincy";
  }
  if (HAVE_FEATURE(get_up_osd_features(), SERVER_PACIFIC) &&
      requested < ceph_release_t::pacific) {
    return "pacific";
  }
  if (HAVE_FEATURE(get_up_osd_features(), SERVER_OCTOPUS) &&
      requested < ceph_release_t::octopus) {
    return "octopus";
  }
  if (HAVE_FEATURE(get_up_osd_features(), SERVER_PACIFIC) &&
      requested < ceph_release_t::pacific) {
    return "pacific";
  }

  return std::nullopt;
}


...
...

  if (auto is_unrequested = unrequested_release(required_osd_release); is_requested) {

    ....*is_requested...

  }

...
...

```

Reply:
@ronen-fr Thanks for the suggestion. I have modified the code accordingly. Since 'require_osd_release' is a part of the OSDMap class, I have not passed it to the suggested method. The rest of the logic remains the same as per your suggestion. Also, this needs back-porting all the way back to nautilus. I suspect for nautilus the logic needs to be changed due to a lower version of C++ being used. Probably boost::optional would need to be employed there.